node/service.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package node
+
+import (
+	"path/filepath"
+	"reflect"
+
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/p2p"
+)
+
+// ServiceContext is a collection of service independent options inherited from
+// the protocol stack, that is passed to all constructors to be optionally used;
+// as well as utility methods to operate on the service environment.
+type ServiceContext struct {
+	datadir  string             // Data directory for protocol persistence
+	services map[string]Service // Index of the already constructed services
+	EventMux *event.TypeMux     // Event multiplexer used for decoupled notifications
+}
+
+// Database opens an existing database with the given name (or creates one if no
+// previous can be found) from within the node's data directory. If the node is
+// an ephemeral one, a memory database is returned.
+func (ctx *ServiceContext) Database(name string, cache int) (ethdb.Database, error) {
```

Comment:
I'd expect this to return **a** database. Not a new one. How about `NewDatabase`?